File: Week 01/id_037/视频中实战题目练习/two_sum.java
// https://leetcode-cn.com/problems/two-sum/
// 暴力法 遍历！
class Solution {
    public int[] twoSum(int[] nums, int target) {
        for (int i = 0; i < nums.length - 1; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                if ( target == nums[i] + nums[j]) {
                    return new int[] {i, j};
                }
            }
        }
        throw new IllegalArgumentException("no results found");
    }
}

// 双指针，从数组两端开始夹逼
// 当双指针的和小于target时，更新i 与 j 所对应的元素的小的一方的指针
// 当双指针的和大于target时，更新i 与 j 所对应的元素的大的一方的指针

// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!
// todo: 解法错误，有待改正!!!!!!!!
class Solution2 {
    public int[] twoSum(int[] nums, int target) {
        int i = 0;
        int j = nums.length - 1;
        while (i != j) {
            if (target > nums[i] + nums[j]) {
                if (nums[i] < nums[j]) {
                    i++;
                } else {
                    j--;
                }
            } else if (target < nums[i] + nums[j]) {
                if (nums[i] < nums[j]) {
                    j--;
                } else {
                    i++;
                }
            } else {
                return new int[] {i, j};
            }
        }
        throw new IllegalArgumentException("no results found");
    }
}

File: Week 01/id_277/two-sum.py
class Solution:
    """
    采用Hash表的方法处理;
    通过构建一个新的字典表,将原本数组中的下标与值分别作为字典的value和Key
    再判断target与循环值的差值是否存在于字典中,从而得出结果
    """
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        dict = {}
        for i, n in enumerate(nums):
            if target - n in dict:
                return [dict[target - n], i]
            dict[n] = i